### PR TITLE
feat: Verify issue planned end time and issue state with iteration

### PR DESCRIPTION
**Pull request by shuofan**
#### What this PR does / why we need it:
Verify issue planned end time and issue state with iteration

#### Which issue(s) this PR fixes:

- Fixes #your-issue_number
- [Erda Cloud Issue Link](https://erda.cloud/erda/dop/projects/387/issues/gantt?filter__urlQuery=eyJpdGVyYXRpb25JRHMiOlsxMTMxLDExMTUsMTE0NiwxMDkwLDExNjRdLCJhc3NpZ25lZSI6WyIxMDAxMDczIl19&id=292854&iterationID=1146&pId=0&type=TASK)

#### Specified Reviewers:

/assign @sfwn 

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |   Verify issue planned end time and issue state with iteration           |
| 🇨🇳 中文    |     事项截止时间和状态与迭代校验         |

#### Need cherry-pick to release versions?

Add comment like `/cherry-pick release/1.0` when this PR is merged.

> For details on the cherry pick process, see the [cherry pick requests](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md#how-to-cherry-pick-a-merged-pr) section under [CONTRIBUTING.md](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by Effet — CHANGES_REQUESTED** [review]

File: modules/dop/services/issue/issue.go [hunk 1]
```diff
@@ -139,6 +139,12 @@ func (svc *Issue) Create(req *apistructs.IssueCreateRequest) (*dao.Issue, error)
 	if req.Creator != "" {
 		req.UserID = req.Creator
 	}
+
+	if req.Type == apistructs.IssueTypeBug || req.Type == apistructs.IssueTypeTask {
+		if err := svc.issueTimeIterationValidator(req.PlanFinishedAt.Value(), req.IterationID); err != nil {
```
[the remark this answers is not in the document]
Comment (shuofan) [reply modules/dop/services/issue/issue.go]: Done. Iteration or state query is called at most once

File: modules/dop/services/issue/adjust.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+	planFinished(condition, *issueValidator) *time.Time
+}
+
+type issueCreateAdjuster struct{}
+
+func (i *issueCreateAdjuster) planFinished(match condition, v *issueValidator) *time.Time {
+	if !match() {
+		return nil
+	}
+	now := time.Date(time.Now().Year(), time.Now().Month(), time.Now().Day(), 0, 0, 0, 0, time.Now().Location())
```
Comment (Effet) [inline modules/dop/services/issue/adjust.go:33]: time.Now() to a temp variable, it call too many. each call will have some offset.

File: modules/dop/services/issue/issue.go [hunk 3]
```diff
@@ -139,6 +139,21 @@ func (svc *Issue) Create(req *apistructs.IssueCreateRequest) (*dao.Issue, error)
 	if req.Creator != "" {
 		req.UserID = req.Creator
 	}
+	if req.Type == apistructs.IssueTypeBug || req.Type == apistructs.IssueTypeTask {
```
Comment (Effet) [inline modules/dop/services/issue/issue.go:142]: can you separate to a func?

File: modules/dop/services/issue/issue.go [hunk 4]
```diff
@@ -621,6 +640,28 @@ func (svc *Issue) UpdateIssue(req apistructs.IssueUpdateRequest) error {
 		issueStreamFields[field] = []interface{}{canUpdateFields[field], v}
 	}
 
+	if issueModel.Type == apistructs.IssueTypeBug || issueModel.Type == apistructs.IssueTypeTask {
```
Comment (Effet) [inline modules/dop/services/issue/issue.go:643]: can you separate to a func?

File: modules/dop/services/issue/issue.go [hunk 5]
```diff
@@ -621,6 +640,28 @@ func (svc *Issue) UpdateIssue(req apistructs.IssueUpdateRequest) error {
...
+		if _, ok := changedFields["plan_finished_at"]; ok {
+			if _, err := validator.TryGetIteration(*req.IterationID); err != nil {
+				return apierrors.ErrUpdateIssue.InternalError(err)
+			}
+			if err := validator.validateIteration(req.PlanFinishedAt.Value()); err != nil {
```
Comment (Effet) [inline modules/dop/services/issue/issue.go:648]: will `req.PlanFinishedAt.Value()` cause panic?

File: modules/dop/services/issue/validate.go [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+	"time"
+
+	"github.com/erda-project/erda/apistructs"
+	"github.com/erda-project/erda/modules/dop/dao"
+)
+
+type issueValidator struct {
+	iteration *dao.Iteration
+	state     *dao.IssueState
+	db        *dao.DBClient
```
Comment (Effet) [inline modules/dop/services/issue/validate.go:28]: it quite like `cache`, can you separate `cache` logic out of validator?
abstract a cache-flow.

File: modules/dop/services/issue/validate.go [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+func (v *issueValidator) TryGetIteration(iterationID int64) (*dao.Iteration, error) {
+	if iterationID <= 0 {
+		return nil, nil
+	}
+	if v.iteration != nil {
+		return v.iteration, nil
+	}
+	iteration, err := v.db.GetIteration(uint64(iterationID))
+	if err != nil {
+		return nil, fmt.Errorf("failed to get iteration, err: %v", err)
```
Comment (Effet) [inline modules/dop/services/issue/validate.go:59]: as a `cache`, if it error every time, should we cache the errors?

**Review by Effet — CHANGES_REQUESTED** [review]

File: modules/dop/services/issue/issue.go (hunk 5 above)
Comment (shuofan) [reply modules/dop/services/issue/issue.go:648]: no, req.PlanFinishedAt is type of IssueTime

File: modules/dop/services/issue/adjust.go (hunk 2 above)
Comment (shuofan) [reply modules/dop/services/issue/adjust.go]: done

File: modules/dop/services/issue/validate.go (hunk 6 above)
Comment (shuofan) [reply modules/dop/services/issue/validate.go]: Changed to use in memory cache

File: modules/dop/services/issue/validate.go (hunk 7 above)
Comment (shuofan) [reply modules/dop/services/issue/validate.go]: No, we cache value instead of err

File: modules/dop/services/issue/issue.go (hunk 3 above)
Comment (shuofan) [reply modules/dop/services/issue/issue.go]: done

File: modules/dop/services/issue/issue.go (hunk 4 above)
Comment (shuofan) [reply modules/dop/services/issue/issue.go:633]: done

File: modules/dop/services/issue/cache.go [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+func (v *issueCache) TryGetIteration(iterationID int64) (*dao.Iteration, error) {
+	if iterationID <= 0 {
+		return nil, nil
+	}
+	iteration, ok := v.iterationCache.LoadWithUpdate(uint64(iterationID))
+	if !ok {
+		return nil, fmt.Errorf("failed to get iteration")
+	}
+	return iteration.(*dao.Iteration), nil
```
Comment (Effet) [inline modules/dop/services/issue/cache.go:58]: will type conversion fail if `iteration` or `state` is nil? I suggest to add some unit test.

File: modules/dop/services/issue/validate.go [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+			c.iteration.Title, c.iteration.StartedAt.Format("2006-01-02"), c.iteration.FinishedAt.Format("2006-01-02"))
+	}
+	return nil
+}
+
+func inIterationInterval(iteration *dao.Iteration, time *time.Time) bool {
+	if time == nil || iteration == nil {
+		return false
+	}
+	return time.After(*iteration.StartedAt) && time.Before(*iteration.FinishedAt)
```
Comment (Effet) [inline modules/dop/services/issue/validate.go:51]: how about the case time == startedAt or item == finishedAt ?

**Review by Effet — CHANGES_REQUESTED** [review]

File: modules/dop/services/issue/cache.go (hunk 8 above)
Comment (shuofan) [reply modules/dop/services/issue/cache.go:58]: add type assertion err

File: modules/dop/services/issue/validate.go (hunk 9 above)
Comment (shuofan) [reply modules/dop/services/issue/validate.go:51]: changed to inclusive range

**Review by Effet — APPROVED** [review]

**Pull request merged**